src/PIL/HeifImagePlugin.py

```diff
@@ -0,0 +1,36 @@
+import io
+
+from . import Image, ImageFile
+
+try:
+    import pyheif
+except ImportError:
+    pyheif = None
+
+def _accept(prefix):
+    check = pyheif.check(prefix)
+    return check != pyheif.heif_filetype_no
+
+
+class HeifImageFile(ImageFile.ImageFile):
+    format = "HEIF"
+    format_description = "HEIF Image File Format"
+
+    def _open(self):
+        b = self.fp.read()
+        if not _accept(b):
+            raise ValueError("Not a HEIF image file")
+
+        heif_file = pyheif.read(b)
+        self._size = heif_file.size
+        self.mode = heif_file.mode
+        self.fp = io.BytesIO(heif_file.data)
```

Comment:
The contract on _open is that we don't do expensive operations, so decoding the entire images and storing a copy violates that. 

Unfortunately, it doesn't appear that pyheif makes it easy to get the metadata that we require for the _open command without decoding the entire image.

Reply:
Thanks, I thought I did have a metadata-only call but maybe I'm not using it. I'll check it out.